### ISSUE TITLE
[courier] esmtpd permissions issue

### ISSUE THREAD

**Issue by markc** [oerdnj/deb.sury.org#203]
I think, still trying to track this one down but in or out SMTP access is not working. IMAP is okay.

Ondřej, is it possible there is some hardwired reference to the "daemon" user when you build the packages, especially for esmtpd?

This is a message to the courier-users mailing-list...
```
I know this is not strictly a courier code issue but I'm trying to track
down a permissions problem and from this strace I am getting...

setgid(1)                               = 0
getuid()                                = 0
setgroups(1, [1])                       = 0
setuid(1)                               = 0
chdir("/usr")                           = 0
chdir("/var/lib/courier/tmp")           = 0
[... ending with ...]
sendto(3, "<19>Jan  2 19:05:47 submit: Permission denied", 45, MSG_NOSIGNAL, NULL, 0) = 45
unlink("145172/1451725547.16796.goldcoast.org") = -1 EACCES (Permission denied)

but there is nothing in /var/lib/courier/tmp.

My question; does that setgid(1) and setuid(1) mean to change to the
uid:gid of 1:1?

That is daemon:daemon on Debian systems whereas it should be 114:117
on my system for the "courier" user. If so then it would explain why
/var/lib/courier/tmp owned by courier:courier can't be written to by
a program changing to daemon:daemon.

Which probably means the packages I am testing are built with some kind
of internal daemon:daemon permissions whereas all external files are
given courier:courier, ie;...

./esmtpd:MAILGROUP=courier
./esmtpd:MAILUSER=courier
```

**Comment by oerdnj** [comment]
What binary program are you stracing?

**Comment by markc** [comment]
Doesn't matter, mail and sendmail both give similar errors when using strace -f. But specifically, the above came from...
```
 mv /usr/lib/courier/courier/submit /usr/lib/courier/courier/submit.orig
chmod +x /usr/lib/courier/courier/submit
cat /usr/lib/courier/courier/submit
 #! /bin/sh

strace -s 256 -o /tmp/submit.`whoami`.$$.log \
     /usr/lib/courier/courier/submit.orig "$@" && \
     rm -f /tmp/submit.`whoami`.$$.log && exit 0
rc=$?
mv /tmp/submit.`whoami`.$$.log /tmp/submit.`whoami`.log.save
exit $?
```

**Comment by markc** [comment]
As I suspected...
```
-------- Forwarded Message --------
Subject: Re: [courier-users] setgid(1) and setuid(1)
Date: Sat, 2 Jan 2016 02:12:28 -0800
From: Gordon Messmer <gordon.messmer@gmail.com>
To: courier-users@lists.sourceforge.net

On 01/02/2016 01:22 AM, Mark Constable wrote:
> My question; does that setgid(1) and setuid(1) mean to change to the
> uid:gid of 1:1?

Yes.

> Which probably means the packages I am testing are built with some kind
> of internal daemon:daemon permissions whereas all external files are
> given courier:courier, ie;...
>
> ./esmtpd:MAILGROUP=courier
> ./esmtpd:MAILUSER=courier

That's correct. The uid and gid that Courier will use are set at compile time.
```

**Comment by markc** [comment]
I know where you are coming from with the `daemon` to `courier` uid:gid change but the simplest solution is to drop the quest for an independent courier-only uid:gid and go back to daemon:daemon as per the original debs.

**Comment by markc** [comment]
FWIW a response from Gordon with a kind of related question (v0.75 introduced $IMAPACCESS and makeimapaccess which I was unaware of)...
```
-------- Forwarded Message --------
Subject: Re: [courier-users] What config option controls -access=
Date: Sat, 2 Jan 2016 02:10:20 -0800
From: Gordon Messmer <gordon.messmer@gmail.com>
To: Courier Users <courier-users@lists.sourceforge.net>

On 01/01/2016 07:48 PM, Mark Constable wrote:
> These new 0.75 deb packages for "wily" (15.10) are proving ornery
> because the packager has changed uid:gid from the previous "daemon"
> user that Debian has used for a decade to the more common "courier"
> user (in line with Archlinux and maybe RH/Centos). IMAP is working
> okay but I'm seeing strange behavior with esmtpd.

Yeah, switching UIDs is not to be taken lightly. It will probably break
working installations, and I tend to think that sort of thing drives
users away.

If the maintainer is intent on this, you'll need to include a
post-install script that fixes the uid and gid of everything that the
various pieces of courier have created, including files in /etc/courier,
/usr/lib/courier, and /var/spool/courier, or whatever layout is used on
Debian.

And for what it's worth, the rpm packages use "daemon" unless there's a
courier user present in the build environment. If you use mock (and you
should), that means that "daemon" is the standard user.
```

**Comment by oerdnj** [comment]
> I know where you are coming from with the daemon to courier uid:gid change but the simplest solution is to drop the quest for an independent courier-only uid:gid and go back to daemon:daemon as per the original debs.

Nah, it's not that hard to dynamically resolve UID and GID on runtime

(And there was also inverted login in postinst script that prevented fixing /var/lib/courier and /etc/courier permissions. Also fixed now locally.)

**Comment by oerdnj** [comment]
Should be hopefully fixed in `courier - 0.75.0-7+deb.sury.org~<dist>+2`. I don't have courier running anywhere, so the testing is up to you.

**Comment by markc** [comment]
Finally I am sending and receiving mails. The main problem, and probably all along aside from some permissions adjustment, was that the `courier` daemon was not being started! The original /etc/init.d/courier-mta startup script starts /usr/sbin/courier, /usr/sbin/courierfilter, /usr/sbin/esmtpd and then /usr/sbin/esmtpd-msa where the last 2 are conditional depending on $ESMTPDSTART in the /etc/courier config files.

It seems that starting multiple daemons from a single systemd service file may not be possible so there will have to be separate files and maybe some adjustment of Requires= and After= may work. I have no idea how to emulate the $ESMTPDSTART env var trick to decide whether to start esmtpd and esmtpd-msa.

It's also possible that the suid programs need to be 4755 root:courier rather than courier:courier but I haven't had time to revert my manual changes or do a fresh install to confirm this.

One extra thing that is needed is to create a /etc/courier/imapaccess directory and to run makeimapaccess whereever makehosteddomains and makeacceptmailfor are run in the install script otherwise imapd/imapd-ssl won't start up.

And /var/run/courier needs to be -g+rwx :courier for the courierfilter.pid file.
```
~ cat /lib/systemd/system/courier.service
[Unit]
Description=Courier Daemon
Requires=courier-authdaemon.service
After=courier-authdaemon.service

[Service]
Type=forking
ExecStart=/usr/sbin/courier start
ExecStop=/usr/sbin/courier stop

[Install]
WantedBy=multi-user.target
```
```
~ cat /lib/systemd/system/courier-filter.service
[Unit]
Description=Courier Filter Daemon
Requires=courier-authdaemon.service
After=courier-authdaemon.service

[Service]
Type=forking
ExecStart=/usr/sbin/courierfilter start
ExecStop=/usr/sbin/courierfilter stop
PIDFile=/var/run/courier/courierfilter.pid

[Install]
WantedBy=multi-user.target
```
